studio/app/optinist/core/edit_ROI/edit_ROI.py

```diff
@@ -1,138 +1,225 @@
+from studio.app.optinist.dataclass import EditRoiData, FluoData, IscellData, RoiData
+from studio.app.optinist.schemas.roi import RoiStatus
 
 
 @dataclass
-class ACTION:
-    EDIT_ROI: str = "edit_ROI"
-    ADD: str = "add"
-    MERGE: str = "merge"
-    DELETE: str = "delete"
+class CellType:
+    ROI = 1
+    NON_ROI = 0
+    TEMP_ADD = -1
+    TEMP_DELETE = -2
 
 
 class EditROI:
-    def __init__(self, action: str, filepath: str, params: dict) -> None:
-        self.action = action
-        self.filepath = filepath
-        self.params = params
-        self.algo = self.get_algo()
-        self.node_dirpath = os.path.dirname(self.filepath)
-
-    def get_algo(self):
-        filepath = self.filepath
-
-        if (
-            not os.path.exists(filepath)
-            and os.path.commonpath([DIRPATH.DATA_DIR, filepath]) != DIRPATH.DATA_DIR
-        ):
-            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND)
-        if os.path.basename(filepath) != "cell_roi.json":
-            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST)
-
-        algo_list = edit_roi_wrapper_dict.keys()
-
-        algo = next((algo for algo in algo_list if algo in filepath), None)
-        if not algo:
-            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST)
-        return algo
-
-    def excute(self):
-        snakemake(
-            DIRPATH.SNAKEMAKE_FILEPATH,
-            use_conda=True,
-            cores=2,
-            workdir=f"{os.path.dirname(DIRPATH.STUDIO_DIR)}",
-            config={
-                "type": ACTION.EDIT_ROI,
-                "action": self.action,
-                "node_dirpath": self.node_dirpath,
-                "algo": self.algo,
-                "params": self.params,
-            },
-        )
+    def __init__(self, file_path):
+        self.node_dirpath = os.path.dirname(file_path)
+
+        self.output_info = self.__read_output_info()
+        self.data: EditRoiData = self.output_info.get("edit_roi_data")
+        self.iscell = self.output_info.get("iscell").data
+        self.fluorescence = self.output_info.get("fluorescence").data
+
+        print("start edit roi:", self.function_id)
 
+    @property
+    def function_id(self):
+        return self.node_dirpath.split("/")[-1]
 
-class EditRoiUtils:
-    @classmethod
-    def conda(cls, config):
-        algo = config["algo"]
-        if "conda_name" in edit_roi_wrapper_dict[algo]:
-            conda_name = edit_roi_wrapper_dict[algo]["conda_name"]
-            return find_condaenv_filepath(conda_name) if conda_name else None
+    @property
+    def pickle_file_path(self):
+        return glob(join_filepath([self.node_dirpath, "*.pkl"]))[0]
 
-        return None
+    @property
+    def shape(self):
+        return self.data.im.shape[1:]
+
+    @property
+    def num_cell(self):
+        return self.data.im.shape[0]
+
+    def get_status(self) -> RoiStatus:
+        return RoiStatus(
+            temp_add_roi=self.data.temp_add_roi,
```

Comment:
@quanpython 
When ROI Edit is executed against ROI data in the past Workflow Record, the following error occurs. (No problem in the ROI immediately after RUN ALL)

> File "D:\workspace\barebone-studio\studio\app\optinist\core\edit_ROI\edit_ROI.py", line 57, in get_status
>     temp_add_roi=self.data.temp_add_roi,
> AttributeError: 'NoneType' object has no attribute 'temp_add_roi'

We are assuming that the requirements require that this ROI Edit be available for past ROIs, so please investigate and address this issue.

Reply:
@itutu-tienday
Because the input data for the ROI editing process has been changed, the old records won't work.